### PR TITLE
ARROW-5757: [Python] Remove Python 2.7 support

### PR DESCRIPTION
**Pull request by pitrou**
Part of the changes were done using [pyupgrade](https://github.com/asottile/pyupgrade).

### CODE REVIEW

**Comment by pitrou** [comment]
Of course, the Python 2.7 build will have to be removed from buildbot.

**Comment by pitrou** [comment]
@ursabot crossbow submit -g wheel

**Comment by pitrou** [comment]
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/30747588

**Comment by pitrou** [comment]
I've checked that `docker-compose run --rm conda-python-hdfs` passes.

**Comment by kszucs** [comment]
@pitrou I consumed my GitHub API rate limit, which affects the crossbow command. I'm going to resubmit it.

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -2157,7 +2151,7 @@ cdef DataType ensure_type(object ty, c_bool allow_none=False):
         return None
     elif isinstance(ty, DataType):
         return ty
-    elif isinstance(ty, six.string_types):
+    elif isinstance(ty, str):
```
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi]: here, string_types is replaces with only `str`, but just above (in the diff), it is replaced with `(bytes, str)`. Is there a reason for the difference?

Comment (pitrou) [reply python/pyarrow/types.pxi]: It seems to me that passing datatypes as bytestrings shouldn't be accepted. As for the other instances in this file, those are about other types, and I wasn't sure so preferred to be conservative.

File: python/setup.py [hunk 2]
```diff
@@ -597,8 +595,7 @@ def has_ext_modules(foo):
     },
     setup_requires=['setuptools_scm', 'cython >= 0.29'] + setup_requires,
     install_requires=install_requires,
-    tests_require=['pytest', 'pandas', 'hypothesis',
-                   'pathlib2; python_version < "3.4"'],
+    tests_require=['pytest', 'pandas', 'hypothesis'],
     description='Python library for Apache Arrow',
     long_description=long_description,
     long_description_content_type='text/markdown',
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/setup.py]: Will do.

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/types.pxi]: Sounds good

File: python/pyarrow/hdfs.py [hunk 3]
```diff
@@ -53,15 +52,15 @@ def _isfilestore(self):
 
     @implements(FileSystem.isdir)
     def isdir(self, path):
-        return super(HadoopFileSystem, self).isdir(path)
+        return super().isdir(path)
```
Comment (wesm) [inline python/pyarrow/hdfs.py]: Oh wow I'm so happy about this.

**Review by wesm — APPROVED** [review]
+1. This was very satisfying to read, thank you for doing this

**Comment by wesm** [comment]
Going to go ahead and merge this

**Pull request closed without merging**